base/multidimensional.jl

```diff
@@ -71,76 +61,83 @@ stagedfunction setindex!{N}(A::AbstractArray, v, index::CartesianIndex{N})
     :((@nref $N A d->getfield(index,d)) = v)
 end
 
-# iteration
-eachindex(A::AbstractArray) = IndexIterator(size(A))
-
-stagedfunction start{N}(iter::IndexIterator{N})
-    indextype, _ = gen_cartesian(N)
-    args = fill(1, N)
-    fieldnames = [symbol("I_$i") for i = 1:N]
-    anyzero = Expr(:(||), [:(iter.dims.$(fieldnames[i]) == 0) for i = 1:N]...)
-    quote
-        z = $anyzero
-        return z, $indextype($(args...))
+# arithmetic, min/max
+for op in (:+, :-, :min, :max)
+    @eval begin
+        stagedfunction ($op){N}(I1::CartesianIndex{N}, I2::CartesianIndex{N})
+            args = [:($($op)(getfield(I1, $d),getfield(I2, $d))) for d = 1:N]
+            :(CartesianIndex{$N}($(args...)))
+        end
     end
 end
 
+stagedfunction *{N}(a::Integer, index::CartesianIndex{N})
+    args = [:(a*getfield(index, $d)) for d = 1:N]
+    :(CartesianIndex{$N}($(args...)))
+end
+*(index::CartesianIndex,a::Integer)=*(a,index)
+
+# Iteration
+immutable CartesianRange{I<:CartesianIndex}
+    start::I
+    stop::I
+end
+
+stagedfunction eachindex{T,N}(A::AbstractArray{T,N})
+    startargs = fill(1, N)
+    stopargs = [:(size(A,$i)) for i=1:N]
+    :(CartesianRange(CartesianIndex{$N}($(startargs...)), CartesianIndex{$N}($(stopargs...))))
+end
+
+stagedfunction start{I}(iter::CartesianRange{I})
+    N=length(I)
+    finishedex = Expr(:(||), [:(iter.stop[$i] < iter.start[$i]) for i = 1:N]...)
+    :(return $finishedex, iter.start)
+end
+
 stagedfunction _start{T,N}(A::AbstractArray{T,N}, ::LinearSlow)
-    indextype, _ = gen_cartesian(N)
     args = fill(1, N)
-    quote
-        z = isempty(A)
-        return z, $indextype($(args...))
-    end
+    :(return isempty(A), CartesianIndex{$N}($(args...)))
 end
 
 # Prevent an ambiguity warning
-gen_cartesian(1) # to make sure the next two lines are valid
-next(R::StepRange, state::(Bool, CartesianIndex{1})) = R[state[2].I_1], (state[2].I_1==length(R), CartesianIndex_1(state[2].I_1+1))
-next{T}(R::UnitRange{T}, state::(Bool, CartesianIndex{1})) = R[state[2].I_1], (state[2].I_1==length(R), CartesianIndex_1(state[2].I_1+1))
+next(R::StepRange, state::(Bool, CartesianIndex{1})) = (index=state[2]; return R[index], (index[1]==length(R), CartesianIndex{1}(index[1]+1)))
```

Comment:
I guess I took the objective of expunging `gen_cartesian` calls to the extreme. However, both `code_typed` and actually calling this and timing it shows that the current version works as intended, i.e. no allocation overhead. However, it is even more important to keep in mind that these methods are only there to prevent warnings, they will never be called since ranges always have efficient linear indexing and will thus not be iterated over in a `::LinearSlow` way.

The benefit of the current approach is that, if we ever have a different implementation of `CartesianIndex`, e.g. because staged types become available, or because it will one day be possible to properly align `NTuple{N,Int}` in an immutable, or there will be a built-in FixedVector type, there will be no need to change any of the iterator code. I guess this was my design goal.